src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -794,25 +790,35 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                     builder.ElementTypeWithAnnotations = collectionType.TypeArgumentsWithAnnotationsNoUseSiteDiagnostics.Single();
 
                     MethodSymbol getEnumeratorMethod = GetGetEnumeratorMethod(collectionType, diagnostics, isAsync, errorLocationSyntax);
+                    MethodSymbol moveNextMethod = null;
 
                     if ((object)getEnumeratorMethod != null)
                     {
                         MethodSymbol specificGetEnumeratorMethod = getEnumeratorMethod.AsMember(collectionType);
-                        builder.GetEnumeratorMethod = specificGetEnumeratorMethod;
                         TypeSymbol enumeratorType = specificGetEnumeratorMethod.ReturnType;
 
                         if (isAsync)
                         {
                             Debug.Assert(enumeratorType.OriginalDefinition.Equals(Compilation.GetWellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)));
+                            Debug.Assert(specificGetEnumeratorMethod.ParameterCount == 1);
+                            Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional);
```

Comment:
>Debug.Assert(specificGetEnumeratorMethod.Parameters[0].IsOptional); [](start = 28, length = 67)

I think it is possible to create a test scenario when this parameter is not declared as optional. Let's add a unit-test and observe behavior.

Reply:
I've created one, we don't handle it correctly today either. Will be fixed in the next commit.

---
In reply to: [549859342](https://github.com/dotnet/roslyn/pull/50105#discussion_r549859342) [](ancestors = 549859342)